src/transformers/models/gpt2/modeling_flax_gpt2.py

```diff
@@ -0,0 +1,645 @@
+            outputs, past_key_values = outputs
+            outputs["past_key_values"] = unfreeze(past_key_values["cache"])
+            return outputs
+        elif past_key_values is not None and not return_dict:
+            outputs, past_key_values = outputs
+            outputs = outputs[:1] + (unfreeze(past_key_values["cache"]),) + outputs[1:]
+
+        return outputs
+
+
+class FlaxGPT2BlockCollection(nn.Module):
+    config: GPT2Config
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.blocks = [
+            FlaxGPT2Block(self.config, name=str(i), dtype=self.dtype) for i in range(self.config.num_hidden_layers)
+        ]
+
+    def __call__(
+        self,
+        hidden_states,
+        attention_mask=None,
+        deterministic: bool = True,
+        _use_cache: bool = False,
+        output_attentions: bool = False,
+        output_hidden_states: bool = False,
+        return_dict: bool = True,
+    ):
+        all_attentions = () if output_attentions else None
+        all_hidden_states = () if output_hidden_states else None
+
+        for block in self.blocks:
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+
+            layer_outputs = block(hidden_states, attention_mask, deterministic=deterministic, _use_cache=_use_cache)
+            hidden_states = layer_outputs[0]
+
+            if output_attentions:
+                all_attentions += (layer_outputs[1],)
+
+        if output_hidden_states:
+            all_hidden_states += (hidden_states,)
+
+        outputs = (hidden_states,)
+
+        if not return_dict:
+            return tuple(v for v in outputs if v is not None)
+
+        return FlaxBaseModelOutputWithPast(
+            last_hidden_state=hidden_states,
+            past_key_values=None,
+            hidden_states=all_hidden_states,
+            attentions=all_attentions,
+        )
+
+
+class FlaxGPT2Module(nn.Module):
+    config: GPT2Config
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.embed_dim = self.config.hidden_size
+
+        self.wte = nn.Embed(
+            self.config.vocab_size,
+            self.embed_dim,
+            embedding_init=jax.nn.initializers.normal(stddev=self.config.initializer_range),
+            dtype=self.dtype,
+        )
+        self.wpe = nn.Embed(
+            self.config.max_position_embeddings,
+            self.embed_dim,
+            embedding_init=jax.nn.initializers.normal(stddev=self.config.initializer_range),
+            dtype=self.dtype,
+        )
+        self.dropout = nn.Dropout(rate=self.config.embd_pdrop)
+        self.h = FlaxGPT2BlockCollection(self.config, dtype=self.dtype)
+        self.ln_f = nn.LayerNorm(epsilon=self.config.layer_norm_epsilon, dtype=self.dtype)
+
+    def __call__(
+        self,
+        input_ids,
+        attention_mask,
+        position_ids,
+        deterministic=True,
+        _use_cache: bool = False,
```

Comment:
Why is this private in the main API when it's public in PyTorch and TensorFlow?

Reply:
I think in Flax the user should never pass `use_cache` but handle everything over:

1) `init_cache(...)` and 
2) passing `past_key_values` or not

It's not possible to pass `use_cache` in the first pass to retrieve the cache like we do it in PyTorch, but we need to call `init_cache(...)` first. => I think it's better to afterward handle everything over `past_key_values` instead of making `use_cache` public. 
We still need the boolean though so that `init_cache(...)` can tell `FlaxGPT2Attention` to initialize the cache. 

=> Another option would be to instead of passing along `_use_cache`, we set a config parameter of `FlaxGPT2Attention` called `is_decoder` to `True` when `past_key_values` and to `False` when it is not instead of passing down a boolean flag & then have no `use_cache` at all? Is this maybe better? @sgugger

Reply:
Actually that won't work either very cleanly since higher modules can't easily access lower modules as in PyTorch. I changed `_use_cache` to `init_cache` now to make a clear differentiation to PyTorch. Users never call `model.module.apply` directly anyways so that should be fine.